File: backend/src/server/gateway.ts
/**
 * TaskFlow AI Gateway pipeline (MASTER_PROJECT_CONTEXT §15).
 *
 *  Authenticate API key → identify customer → identify subscription
 *  → check entitlement → check rate limit → check credit balance
 *  → validate model access → route request → receive response
 *  → calculate usage → record usage event → deduct credits → respond.
 *
 * BILLING SAFETY: usage is only billed when an upstream call succeeded.
 * Nothing is debited for requests that never reached a provider (§15, §29).
 * Customer-supplied usage/pricing values are never trusted (§29).
 */
import { randomBytes } from "crypto";
import { db } from "@/lib/db";
import { authenticateApiKey } from "@/server/auth/api-keys";
import { planEntitled } from "@/server/billing/subscription";
import { checkRateLimit, tryAcquireConcurrency, releaseConcurrency } from "@/server/ratelimit";
import { priceUsage } from "@/server/billing/pricing";
import { recordUsage } from "@/server/usage/recorder";
import { routeModel, isProviderConfigured } from "@/server/providers/registry";
import {
  ProviderChatMessage,
  ProviderError,
} from "@/server/providers/types";
import { MICRO_PER_POUND } from "@/lib/money";

export type GatewayRequestBody = {
  model?: unknown;
  messages?: unknown;
  temperature?: unknown;
  max_tokens?: unknown;
  stream?: unknown;
};

export type GatewayError = {
  status: number;
  code: string;
  message: string;
  retryAfterSeconds?: number;
};

type GatewaySuccess = {
  ok: true;
  response: {
    id: string;
    object: "chat.completion";
    created: number;
    model: string;
    choices: Array<{
      index: number;
      message: { role: "assistant"; content: string };
      finish_reason: string | null;
    }>;
    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
      estimated?: boolean;
    };
    taskflow: {
      request_id: string;
      provider: string;
      charge_micros: number;
      balance_micros: number | null;
      latency_ms: number;
    };
  };
};

type GatewayFailure = { ok: false; error: GatewayError };

export type GatewayResult = GatewaySuccess | GatewayFailure;

function gatewayError(status: number, code: string, message: string, retryAfterSeconds?: number): GatewayFailure {
  return { ok: false, error: { status, code, message, retryAfterSeconds } };
}

/** Minimal OpenAI-shaped error payload. */
export function errorPayload(err: GatewayError) {
  return {
    error: {
      message: err.message,
      type: "taskflow_error",
      code: err.code,
    },
  };
}

function parseMessages(raw: unknown): ProviderChatMessage[] | null {
  if (!Array.isArray(raw) || raw.length === 0) return null;
  const out: ProviderChatMessage[] = [];
  for (const item of raw) {
    if (!item || typeof item !== "object") return null;
    const role = (item as Record<string, unknown>).role;
    const content = (item as Record<string, unknown>).content;
    if (typeof role !== "string" || typeof content !== "string") return null;
    if (!["system", "user", "assistant"].includes(role)) return null;
    out.push({ role: role as ProviderChatMessage["role"], content });
  }
  return out;
}

export async function handleChatCompletion(
  rawKey: string | null,
  body: GatewayRequestBody,
): Promise<GatewayResult> {
  // ── 1. Authenticate API key ────────────────────────────────
  if (!rawKey) {
    return gatewayError(401, "missing_api_key", "Provide your TaskFlow API key as 'Authorization: Bearer tf_live_…'");
  }
  const auth = await authenticateApiKey(rawKey);
  if (!auth.ok) {
    return gatewayError(auth.status, "invalid_api_key", auth.error);
  }
  const { key } = auth;
  const subscription = key.user.subscription;
  const plan = subscription?.plan;

  // ── 2. Subscription entitlement (a key alone is not access — §19) ──
  if (!subscription || !plan) {
    return gatewayError(403, "no_subscription", "Account has no subscription. Subscribe to a plan to use the API.");
  }
  if (!planEntitled(subscription.status, subscription.pastDueAt)) {
    return gatewayError(
      402,
      "subscription_inactive",
      subscription.status === "PAST_DUE"
        ? "Payment is past due and the grace window has ended. Update billing to continue."
        : `Subscription status '${subscription.status}' does not grant API access. Update billing to continue.`,
    );
  }

  // ── 3. Model request validation ────────────────────────────
  if (typeof body.model !== "string" || !body.model) {
    return gatewayError(400, "invalid_request", "'model' is required");
  }
  const messages = parseMessages(body.messages);
  if (!messages) {
    return gatewayError(400, "invalid_request", "'messages' must be an array of {role, content} objects");
  }
  if (body.stream === true) {
    // Streaming metering lands with the streaming gateway; rejected honestly for now.
    return gatewayError(400, "streaming_unsupported", "Streaming is not enabled yet; use non-streaming requests.");
  }
  const temperature = typeof body.temperature === "number" ? body.temperature : undefined;
  const maxTokens = typeof body.max_tokens === "number" ? Math.floor(body.max_tokens) : undefined;

  // ── 4. Entitlement: is this model allowed for the plan? ────
  const entitlement = await db.modelEntitlement.findUnique({
    where: { modelId_planId: { modelId: body.model, planId: plan.id } },
  });
  if (!entitlement) {
    return gatewayError(
      403,
      "model_not_entitled",
      `Model '${body.model}' is not available on the ${plan.name} plan. Upgrade or choose an entitled model.`,
    );
  }

  // ── 5. Rate limit (per-plan RPM) ───────────────────────────
  const rl = checkRateLimit(key.id, plan.rateLimitPerMinute);
  if (!rl.allowed) {
    return gatewayError(
      429,
      "rate_limit_exceeded",
      `Rate limit of ${rl.limit} requests/minute exceeded for the ${plan.name} plan.`,
      rl.retryAfterSeconds,
    );
  }

  // ── 6. Concurrency cap (per-plan) ──────────────────────────
  if (!tryAcquireConcurrency(key.id, plan.maxConcurrency)) {
    return gatewayError(
      429,
      "concurrency_limit",
      `Concurrency limit of ${plan.maxConcurrency} in-flight requests reached for the ${plan.name} plan.`,
    );
  }

  const requestId = `tf-req-${randomBytes(12).toString("hex")}`;
  const started = Date.now();

  try {
    // ── 7. Credit balance check (gate before spending provider cost) ──
    const balance = key.user.creditAccount?.balanceMicros ?? 0;
    if (balance <= 0) {
      return gatewayError(
        402,
        "insufficient_credits",
        "AI credit balance exhausted. Purchase additional credits or wait for the monthly allowance.",
      );
    }

    // ── 8. Route through the provider abstraction ──────────────
    const route = await routeModel(body.model);

    // ── 9. Upstream call ───────────────────────────────────────
    const result = await route.adapter.chatCompletion({
      model: route.upstreamModel,
      messages,
      temperature,
      maxTokens,
    });
    const latencyMs = Date.now() - started;

    // ── 10. Price usage (server-side only — never customer-supplied) ──
    const priced = await priceUsage(route.modelId, {
      inputTokens: result.usage.inputTokens,
      outputTokens: result.usage.outputTokens,
    });

    // Pricing integrity (§43/§64): a model without a rate card must never be
    // served for free. Record an unbilled FAILED event for the audit trail and
    // refuse the response — the upstream call is written off, never under-billed.
    if (!priced.rateCard) {
      await recordUsage({
        requestId,
        userId: key.userId,
        organizationId: key.organizationId,
        apiKeyId: key.id,
        modelId: route.modelId,
        provider: route.provider,
        breakdown: {
          inputTokens: result.usage.inputTokens,
          outputTokens: result.usage.outputTokens,
          totalTokens: result.usage.inputTokens + result.usage.outputTokens,
          computeSeconds: 0,
          providerCostMicros: 0,
          computeCostMicros: 0,
          marginMicros: 0,
          customerChargeMicros: 0,
        },
        status: "FAILED",
        latencyMs,
        errorMessage: "pricing_not_configured: no rate card for model; request not billed",
        metadata: { plan: plan.id, keyPrefix: key.prefix },
        billCustomer: false,
      }).catch(() => undefined);
      return gatewayError(
        500,
        "pricing_not_configured",
        "Model pricing is not configured on this deployment. The request was not served and nothing was charged.",
      );
    }

    // ── 11. Record usage event + deduct credits atomically ─────
    const recorded = await recordUsage({
      requestId,
      userId: key.userId,
      organizationId: key.organizationId,
      apiKeyId: key.id,
      modelId: route.modelId,
      provider: route.provider,
      breakdown: priced.breakdown,
      status: "SUCCESS",
      latencyMs,
      metadata: {
        tokensEstimated: result.usage.estimated,
        rateCard: priced.rateCard,
        plan: plan.id,
        keyPrefix: key.prefix,
      },
      billCustomer: true,
    });

    await db.apiKey.update({
      where: { id: key.id },
      data: { lastUsedAt: new Date() },
    }).catch(() => undefined);

    // ── 12. OpenAI-shaped response with TaskFlow metering block ──
    return {
      ok: true,
      response: {
        id: requestId,
        object: "chat.completion",
        created: Math.floor(Date.now() / 1000),
        model: route.modelId,
        choices: [
          {
            index: 0,
            message: { role: "assistant", content: result.content },
            finish_reason: result.finishReason,
          },
        ],
        usage: {
          prompt_tokens: result.usage.inputTokens,
          completion_tokens: result.usage.outputTokens,
          total_tokens: result.usage.inputTokens + result.usage.outputTokens,
          ...(result.usage.estimated ? { estimated: true } : {}),
        },
        taskflow: {
          request_id: requestId,
          provider: route.provider,
          charge_micros: recorded.chargedMicros,
          balance_micros: recorded.balanceMicros,
          latency_ms: latencyMs,
        },
      },
    };
  } catch (err) {
    const latencyMs = Date.now() - started;

    if (err instanceof ProviderError) {
      // Upstream failed → usage event (FAILED) for observability, NO charge.
      // We record zero-token failure events only for provider-dispatch errors,
      // keeping the audit trail without inventing usage.
      await recordUsage({
        requestId,
        userId: key.userId,
        organizationId: key.organizationId,
        apiKeyId: key.id,
        modelId: body.model,
        provider: err.provider,
        breakdown: {
          inputTokens: 0,
          outputTokens: 0,
          totalTokens: 0,
          computeSeconds: 0,
          providerCostMicros: 0,
          computeCostMicros: 0,
          marginMicros: 0,
          customerChargeMicros: 0,
        },
        status: "FAILED",
        latencyMs,
        errorMessage: err.message,
        metadata: { plan: plan.id, keyPrefix: key.prefix },
        billCustomer: false,
      }).catch(() => undefined);

      return gatewayError(err.status, "provider_error", err.message);
    }

    return gatewayError(500, "gateway_error", "Unexpected gateway error. The request was not billed.");
  } finally {
    releaseConcurrency(key.id);
  }
}

/** OpenAI-style model listing for key holders (entitled + actually usable models). */
export async function handleListModels(rawKey: string | null) {
  if (!rawKey) {
    return gatewayError(401, "missing_api_key", "Provide your TaskFlow API key as 'Authorization: Bearer tf_live_…'");
  }
  const auth = await authenticateApiKey(rawKey);
  if (!auth.ok) return gatewayError(auth.status, "invalid_api_key", auth.error);

  const planId = auth.key.user.subscription?.planId ?? "free";
  const entitlements = await db.modelEntitlement.findMany({
    where: { planId, model: { isActive: true } },
    include: { model: { include: { pricing: true } } },
    orderBy: { model: { sortOrder: "asc" } },
  });

  // Only advertise models that can actually be served right now (§6):
  // registry status LIVE AND provider configured on this deployment.
  const data = entitlements
    .filter((e) => (!e.model.status || e.model.status === "LIVE") && isProviderConfigured(e.model.provider))
    .map((e) => ({
      id: e.model.id,
      object: "model",
      owned_by: e.model.provider,
      created: Math.floor(Date.now() / 1000),
      taskflow: {
        display_name: e.model.displayName,
        capabilities: JSON.parse(e.model.capabilities || "[]"),
        context_window: e.model.contextWindow,
        hosting_mode: e.model.hostingMode,
        status: e.model.status,
        pricing: e.model.pricing
          ? {
              input_per_1k: e.model.pricing.inputCostPer1kMicros / MICRO_PER_POUND,
              output_per_1k: e.model.pricing.outputCostPer1kMicros / MICRO_PER_POUND,
              currency: "GBP",
            }
          : null,
      },
    }));

  return { ok: true as const, data };
}

File: backend/app/api/gateway.py
from __future__ import annotations

import time
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Header, HTTPException, Request
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.database import get_db
from app.auth.maintenance import block_if_maintenance_unconditional
from app.models.api_key import ApiKey
from app.models.billing import CreditTransactionType, Plan, Subscription, SubscriptionStatus
from app.models.catalog import AiModel, ModelEntitlement, ModelStatus, Provider
from app.models.usage import UsageEvent, UsageEventStatus
from app.providers.base import ProviderRequestError
from app.providers.registry import get_provider
from app.services import api_keys as api_key_service
from app.services import credits as credits_service
from app.services import pricing as pricing_service

router = APIRouter(prefix="/v1", tags=["gateway"], dependencies=[Depends(block_if_maintenance_unconditional)])


async def authenticate(authorization: str | None = Header(default=None), db: Session = Depends(get_db)) -> ApiKey:
    if not authorization or not authorization.lower().startswith("bearer "):
        raise HTTPException(401, "Missing API key")
    raw_key = authorization.split(" ", 1)[1]
    key = api_key_service.authenticate_api_key(db, raw_key)
    if not key:
        raise HTTPException(401, "Invalid or revoked API key")
    return key


@router.get("/models")
def list_models(api_key: ApiKey = Depends(authenticate), db: Session = Depends(get_db)):
    sub = db.execute(select(Subscription).where(Subscription.organization_id == api_key.organization_id)).scalar_one_or_none()
    plan = db.get(Plan, sub.plan_id) if sub else None
    plan_code = plan.code if plan else None

    models = db.execute(select(AiModel).where(AiModel.enabled == True, AiModel.status == ModelStatus.LIVE)).scalars()  # noqa: E712
    result = []
    for m in models:
        entitlements = {e.plan_code for e in m.entitlements}
        if plan_code and plan_code not in entitlements:
            continue
        result.append({"id": m.slug, "object": "model", "owned_by": "taskflow"})
    return {"object": "list", "data": result}


class ChatMessage(BaseModel):
    role: str
    content: str


class ChatCompletionRequest(BaseModel):
    model: str
    messages: list[ChatMessage]
    temperature: float | None = None
    max_tokens: int | None = None


@router.post("/chat/completions")
async def chat_completions(
    body: ChatCompletionRequest, api_key: ApiKey = Depends(authenticate), db: Session = Depends(get_db)
):
    org_id = api_key.organization_id

    # 1. Resolve subscription/plan/entitlement
    sub = db.execute(select(Subscription).where(Subscription.organization_id == org_id)).scalar_one_or_none()
    if not sub or sub.status not in (SubscriptionStatus.ACTIVE, SubscriptionStatus.TRIALING):
        raise HTTPException(402, "Organization does not have an active subscription")
    plan = db.get(Plan, sub.plan_id)

    model = db.execute(select(AiModel).where(AiModel.slug == body.model)).scalar_one_or_none()
    if not model or not model.enabled or model.status != ModelStatus.LIVE:
        raise HTTPException(404, f"Model '{body.model}' is not available")

    entitled_plans = {e.plan_code for e in model.entitlements}
    if plan.code not in entitled_plans:
        raise HTTPException(403, f"Model '{body.model}' is not available on the {plan.code.value} plan")

    provider = db.get(Provider, model.provider_id)
    adapter = get_provider(provider.kind.value)
    if not adapter or not adapter.is_configured():
        _record_rejected(db, org_id, model.id, provider.id, "Provider not configured")
        raise HTTPException(503, f"Provider '{provider.kind.value}' is not currently configured")

    # 2. Estimate a conservative pre-authorization charge so we never let a
    #    request through with an empty/negative balance (credit safety, docs §43).
    account = credits_service.get_or_create_account(db, org_id)
    if account.balance_micros <= 0:
        _record_rejected(db, org_id, model.id, provider.id, "Insufficient credits")
        db.commit()
        raise HTTPException(402, "Insufficient credit balance")

    # 3. Dispatch to the provider. Only AFTER this succeeds do we charge —
    #    a request that never reaches the provider is never billed as success.
    try:
        result = await adapter.chat_completion(
            model_identifier=model.model_identifier,
            messages=[m.model_dump() for m in body.messages],
            temperature=body.temperature,
            max_tokens=body.max_tokens,
        )
    except ProviderRequestError as exc:
        _record_provider_error(db, org_id, model.id, provider.id, str(exc))
        db.commit()
        raise HTTPException(exc.status_code, str(exc))

    # 4. Price and charge, transactionally, only on confirmed success.
    priced = pricing_service.price_chat_usage(db, model, result.input_tokens, result.output_tokens)

    event = UsageEvent(
        id=uuid.uuid4(),
        organization_id=org_id,
        api_key_id=api_key.id,
        model_id=model.id,
        provider_id=provider.id,
        status=UsageEventStatus.SUCCESS,
        input_tokens=result.input_tokens,
        output_tokens=result.output_tokens,
        total_tokens=result.input_tokens + result.output_tokens,
        provider_cost_micros=priced.provider_cost_micros,
        gpu_cost_micros=priced.gpu_cost_micros,
        margin_micros=priced.margin_micros,
        charge_micros=priced.charge_micros,
        latency_ms=result.latency_ms,
        created_at=datetime.now(timezone.utc),
    )
    db.add(event)
    db.flush()

    try:
        credits_service.record_transaction(
            db,
            organization_id=org_id,
            type=CreditTransactionType.USAGE,
            amount_micros=-priced.charge_micros,
            description=f"{model.display_name} usage",
            allow_overdraft=True,  # request already ran; never leave an unbilled successful request
            usage_event_id=event.id,
        )
    finally:
        db.commit()

    return {
        "id": f"chatcmpl_{event.id}",
        "object": "chat.completion",
        "model": body.model,
        "choices": [{"index": 0, "message": {"role": "assistant", "content": result.content}, "finish_reason": "stop"}],
        "usage": {
            "prompt_tokens": result.input_tokens,
            "completion_tokens": result.output_tokens,
            "total_tokens": result.input_tokens + result.output_tokens,
        },
        "taskflow": {
            "charge_micros": priced.charge_micros,
            "balance_micros": credits_service.get_or_create_account(db, org_id).balance_micros,
            "provider": provider.kind.value,
        },
    }


def _record_rejected(db: Session, org_id: uuid.UUID, model_id, provider_id, detail: str) -> None:
    db.add(
        UsageEvent(
            id=uuid.uuid4(),
            organization_id=org_id,
            model_id=model_id,
            provider_id=provider_id,
            status=UsageEventStatus.REJECTED,
            error_detail=detail[:500],
            created_at=datetime.now(timezone.utc),
        )
    )


def _record_provider_error(db: Session, org_id: uuid.UUID, model_id, provider_id, detail: str) -> None:
    db.add(
        UsageEvent(
            id=uuid.uuid4(),
            organization_id=org_id,
            model_id=model_id,
            provider_id=provider_id,
            status=UsageEventStatus.PROVIDER_ERROR,
            error_detail=detail[:500],
            created_at=datetime.now(timezone.utc),
        )
    )
